# Python - Easier to learn like normal English language
It is an Interpreter
Executes the code line by line (one statement at a time)


Author : Shanmugaraj Kamaraj

![](Capture1.png)

![](Capture2.png)

###### Print Statement

In [1]:
print("Hello environment")

Hello environment


In [2]:
print("Hello environment", end="")

Hello environment

In [3]:
print("Hello environment", end=" again")

Hello environment again

In [4]:
print("Hello environment","good day", end=" again", sep="*")

Hello environment*good day again

In [5]:
print("Hello environment","good day","Happy day", end=" again", sep=" and ")

Hello environment and good day and Happy day again

###### Comments(#)

In [6]:
# comments are give for the notes about the code, which starts with #

###### Values and Variables

The interpreter, however, can evaluate a Python expression

In [7]:
4

4

In [8]:
"Hello"

'Hello'

In [9]:
'Hello'

'Hello'

In [10]:
"1.02"

'1.02'

In [11]:
2.0

2.0

![](Capture3.png)

In [12]:
2+2

4

In [13]:
2/2

1.0

In [14]:
2//2 

1

In [15]:
2*2 

4

In [16]:
2-2

0

In [17]:
2 % 2

0

In [18]:
-2 ** 3 #It means 2*2*2

-8

In [19]:
print(1 + 2 + 4 + 10 + 3)

20


In [20]:
"fine"

'fine'

In [21]:
fine

NameError: name 'fine' is not defined

The Definition of Name error in the above code -

"fine" (with quotes) and fine (without quotes) are different. 
fine (without quotes) is considere as a variable but "fine"(with quotes) is consider as a string

Type function - returns what type is

In [22]:
type(4)

int

In [23]:
type("4")

str

In [24]:
type(4.0)

float

Convertion in variables types

In [25]:
a = 1024.0
print(type(a), a)
a = int(a)
print(type(a), a)

<class 'float'> 1024.0
<class 'int'> 1024


In [26]:
a = 1024
print(type(a), a)
a = float(a)
print(type(a), a)


<class 'int'> 1024
<class 'float'> 1024.0


In [27]:
a = 1024
print(type(a), a)
a = str(a)
print(type(a), a)

<class 'int'> 1024
<class 'str'> 1024


In [28]:
a = "1024"
print(type(a), a)
a = int(a)
print(type(a), a)

<class 'str'> 1024
<class 'int'> 1024


In [29]:
a = "1024.2"
print(type(a), a)
a = float(a)
print(type(a), a)

<class 'str'> 1024.2
<class 'float'> 1024.2


In [30]:
a = "hello"
print(type(a), a)
a = float(a)
print(type(a), a)

<class 'str'> hello


ValueError: could not convert string to float: 'hello'

In the above code, we try to convert the string "hello" to float. But it is not possible to change the words to numbers using the built in function like int(), str(), float()

In [31]:
a = "hello"
print(type(a))
a = int(a)
print(type(a))

<class 'str'>


ValueError: invalid literal for int() with base 10: 'hello'

In the above code, we try to convert the string "hello" to integer. But it is not possible to change the words to numbers using the built in function like int(), str(), float()

In [32]:
from sys import getsizeof
print(getsizeof(a))
a = a*10000000
print(getsizeof(a))
#getsizeof returns the size in bytes

54
50000049


![](Capture4.png)

In [33]:
5 + 5

10

In [34]:
'5' + '5'

'55'

In [35]:
'hello' + 'py'

'hellopy'

In [36]:
'hello' + ' ' + 'py'

'hello py'

In [37]:
’5’ + 10
#It is makes error because one is integer(10) and another is string('5')

SyntaxError: invalid character in identifier (<ipython-input-37-f67789a39ca3>, line 1)

In [38]:
int('5') + 10
# We can convert the string to integer and then added to the number

15

In [39]:
'5' + str(10)

'510'

In [40]:
type('4' + '4')

str

In [41]:
type('hello' + '4')

str

In [42]:
type('4' + 4)
#Can not concate the num and string

TypeError: can only concatenate str (not "int") to str

![](Capture5.png)

In [43]:
x = 10
print(x)

10


In [44]:
X = "python"
print(X)

python


The variable x and X is different.
Python is case sensitive

In [45]:
x, y, z = 100, -45, 0
print('x =', x, ' y =', y, ' z =', z)

x = 100  y = -45  z = 0


In [46]:
a = 2 
b = 5 
print('a:',a, 'b:',b)
a = b 
print('a:',a, 'b:',b)
b = 7
print('a:',a, 'b:',b)


a: 2 b: 5
a: 5 b: 5
a: 5 b: 7


![](Capture6.png)

![](Capture7.png)

Identifiers or variables must contain at least one character.

The first character must be an alphabetic letter (upper or lower case) or the underscore. 

Like - ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz_ 

There maining characters (if any) may be alphabetic characters (upper or lower case), the underscore, or a digit. 

Like - ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz_0123456789

No other characters (including spaces) are permitted in identifiers or variables

A reserved word or keywords cannot be used as an identifier or variables

In [47]:
class = 15
#class is a reserved word

SyntaxError: invalid syntax (<ipython-input-47-98b9d9a435e2>, line 1)

In [48]:
print('hello')
type(print)

hello


builtin_function_or_method

In [49]:
print = 6
type(print)

int

In [50]:
print

6

In [51]:
print("d")
#Error : because of changeing the print function to an integer
#it generally is not a good idea to do like this

TypeError: 'int' object is not callable

In [52]:
dir() #returns the function and variables in the current directory

['In',
 'Out',
 'X',
 '_',
 '_10',
 '_11',
 '_12',
 '_13',
 '_14',
 '_15',
 '_16',
 '_17',
 '_18',
 '_20',
 '_22',
 '_23',
 '_24',
 '_33',
 '_34',
 '_35',
 '_36',
 '_38',
 '_39',
 '_40',
 '_41',
 '_48',
 '_49',
 '_50',
 '_7',
 '_8',
 '_9',
 '__',
 '___',
 '__builtin__',
 '__builtins__',
 '__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_dh',
 '_i',
 '_i1',
 '_i10',
 '_i11',
 '_i12',
 '_i13',
 '_i14',
 '_i15',
 '_i16',
 '_i17',
 '_i18',
 '_i19',
 '_i2',
 '_i20',
 '_i21',
 '_i22',
 '_i23',
 '_i24',
 '_i25',
 '_i26',
 '_i27',
 '_i28',
 '_i29',
 '_i3',
 '_i30',
 '_i31',
 '_i32',
 '_i33',
 '_i34',
 '_i35',
 '_i36',
 '_i37',
 '_i38',
 '_i39',
 '_i4',
 '_i40',
 '_i41',
 '_i42',
 '_i43',
 '_i44',
 '_i45',
 '_i46',
 '_i47',
 '_i48',
 '_i49',
 '_i5',
 '_i50',
 '_i51',
 '_i52',
 '_i6',
 '_i7',
 '_i8',
 '_i9',
 '_ih',
 '_ii',
 '_iii',
 '_oh',
 'a',
 'b',
 'exit',
 'get_ipython',
 'getsizeof',
 'print',
 'quit',
 'x',
 'y',
 'z']

In [53]:
__builtin__.print("hello again")  # This is the print() function which prints whatever we give it, to the screen or shell

hello again


In [54]:
print("d") # Because, now, print is an integer 

TypeError: 'int' object is not callable

In [55]:
print = __builtin__.print  # This is the print() function which prints whatever we give it, to the screen or shell

In [56]:
print("hello print")

hello print


In [57]:
type(print)

builtin_function_or_method

In [58]:
pi = 23.3123400654033989

In [59]:
pi

23.312340065403397

The float has limited storage size. That is why, it stores the number 23.3123400654033989 as 23.312340065403397

In [60]:
avogadros_number = 6.022e23
c = 2.998e8
print("Avogadro's number =", avogadros_number)
print("Speed of light =", c)

Avogadro's number = 6.022e+23
Speed of light = 299800000.0


![](Capture8.png)

In [61]:
print('A\nB\nC')

A
B
C


In [62]:
print('D\tE\tF')

D	E	F


In [63]:
print('WX\bYZ')

WXYZ


In [64]:
print('1\a2\a3\a4\a5\a6')

123456


In [65]:
print("Did you know that 'word' is a word?")
print('Did you know that "word" is a word?')
print('Did you know that \'word\' is a word?')
print("Did you know that \"word\" is a word?")

Did you know that 'word' is a word?
Did you know that "word" is a word?
Did you know that 'word' is a word?
Did you know that "word" is a word?


In [66]:
filename_path = 'C:\\Users\\rick'
print(filename_path)

C:\Users\rick


![](Capture9.png)

In [67]:
input()

good


'good'

In [68]:
type(input)

method

In [69]:
print('Please enter some text:') 
x = input() 
print('Text entered:', x) 
print('Type:', type(x))

Please enter some text:
good
Text entered: good
Type: <class 'str'>


The input function produces a string from the user’s keyboard input.

In [70]:
print('Please enter some number:') 
x = float(input())
print('Text entered:', x) 
print('Type:', type(x))

Please enter some number:
69
Text entered: 69.0
Type: <class 'float'>


![](Capture10.png)

In [71]:
x1=eval(input('Give Int/Float/String : '))
print('x1=',x1,'type:',type(x1))

Give Int/Float/String : 2
x1= 2 type: <class 'int'>


In [72]:
x1=eval(input('Give Int/Float/String : '))
print('x1=',x1,'type:',type(x1))
#The given string must be in the quotes

Give Int/Float/String : 3.0
x1= 3.0 type: <class 'float'>


In [73]:
num1, num2 = eval(input('Please enter number 1, number 2: '))
print(num1, '+', num2, '=', num1 + num2)

Please enter number 1, number 2: 2,3.0
2 + 3.0 = 5.0


Chained assignment can be used to assign the same value to multiple variables within one statement

In [74]:
a = b = c = 5 #Chained assignment
print(a,b,c)

5 5 5


![](Capture11.png)

In [75]:
True

True

In [76]:
False

False

![](Capture12.png)

In [77]:
if(a==b):
    print("match")
elif(a>b):
    print("a is greater")
else:
    print("else")

match


In [78]:
if(a==b) and (a>b):
    print("match and a is greater")
elif(a==b) or (a>b):
    print("match or a is greater")
if not (a==b):
    print("different")


match or a is greater


In [79]:
count = 1 
while count <= 5:
    print(count) 
    count += 1 #

1
2
3
4
5


In [80]:
for n in range(1, 11): 
    print(n)

1
2
3
4
5
6
7
8
9
10


![](Capture14.png)

In [81]:
lst = [5, -3, 12]
lst

[5, -3, 12]

In [82]:
for i in lst:
    print(i)

5
-3
12


In [83]:
print(lst[0])
print(lst[1])
print(lst[2])

5
-3
12


In [84]:
#List of List
col = [23, [9.3, 11.2, 99.0], [23], [], 4, [0, 0]] 
print(col)

[23, [9.3, 11.2, 99.0], [23], [], 4, [0, 0]]


In [85]:
for i in col:
    print(i)

23
[9.3, 11.2, 99.0]
[23]
[]
4
[0, 0]


In [86]:
a = [10, 20, 30, 40]
b = [10, 20, 30, 40]
print('a =', a)
print('b =', b)
b[2] = 35
print('a =', a) 
print('b =', b)

a = [10, 20, 30, 40]
b = [10, 20, 30, 40]
a = [10, 20, 30, 40]
b = [10, 20, 35, 40]


###### Lists are mutable data structures

In [87]:
a=[10,20,30,40]
b=[10,20,30,40]
print('Is',a,' equal to ',b,'?',sep='',end='')
print()
print(a==b)
print('Are',a,' and ',b,'aliases?',sep='',end='')
print()
print(a is b)
c=[100,200,300,400]
d=c
print('Is',c,' equal to ',d,'?',sep='',end='')
print()
print(c==d)
print('Are',c,' and ',d,'aliases?',sep='',end='')
print()
print(c is d)

Is[10, 20, 30, 40] equal to [10, 20, 30, 40]?
True
Are[10, 20, 30, 40] and [10, 20, 30, 40]aliases?
False
Is[100, 200, 300, 400] equal to [100, 200, 300, 400]?
True
Are[100, 200, 300, 400] and [100, 200, 300, 400]aliases?
True


In [88]:
c=[100,200,300,400]
d=c
print("c = ",c)
print("d = ",d)
c[0] = 22
print("c = ",c)
print("d = ",d)
d[0] = 0
print("c = ",c)
print("d = ",d)

c =  [100, 200, 300, 400]
d =  [100, 200, 300, 400]
c =  [22, 200, 300, 400]
d =  [22, 200, 300, 400]
c =  [0, 200, 300, 400]
d =  [0, 200, 300, 400]


In [89]:
import copy # This code does not make reference to same object. In this case, object is [100,200,300,400]
c=[100,200,300,400]
d=copy.copy(c)
print("c = ",c)
print("d = ",d)
c[0] = 22
print("c = ",c)
print("d = ",d)
d[0] = 0
print("c = ",c)
print("d = ",d)

c =  [100, 200, 300, 400]
d =  [100, 200, 300, 400]
c =  [22, 200, 300, 400]
d =  [100, 200, 300, 400]
c =  [22, 200, 300, 400]
d =  [0, 200, 300, 400]


In [90]:
data = [10, 20, 30, 40, 50, 60] 
print(data[0]) 
print(data[1]) 
print(data[2]) 
print(data[3]) 
print(data[4]) 
print(data[5]) 
print("--------")
print(data[-1]) 
print(data[-2]) 
print(data[-3]) 
print(data[-4]) 
print(data[-5]) 
print(data[-6])

10
20
30
40
50
60
--------
60
50
40
30
20
10


![](Capture15.png)

list [ begin : end ] # Syntax for slicing List 

In [91]:
lst = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]

In [92]:
type(lst)

list

In [93]:
print(lst)
print(lst[0:3])
print(lst[4:8])
print(lst[2:5])
print(lst[-5:-3])
print(lst[:3])
print(lst[4:])
print(lst[:])
print(lst[-100:3])
print(lst[4:100])

[10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
[10, 20, 30]
[50, 60, 70, 80]
[30, 40, 50]
[60, 70]
[10, 20, 30]
[50, 60, 70, 80, 90, 100]
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
[10, 20, 30]
[50, 60, 70, 80, 90, 100]


In [94]:
lst[::2] #jumps in index 0 to 2 to 4 to 6

[10, 30, 50, 70, 90]

In [95]:
len(lst)

10

In [96]:
lst[0] = "hello"

In [97]:
lst

['hello', 20, 30, 40, 50, 60, 70, 80, 90, 100]

In [98]:
del lst[0]
print(lst)

[20, 30, 40, 50, 60, 70, 80, 90, 100]


In [99]:
lst.append("hello")
print(lst)

[20, 30, 40, 50, 60, 70, 80, 90, 100, 'hello']


In [100]:
'hello' in lst

True

In [101]:
"good" in lst

False

In [102]:
lst.index('hello')

9

In [103]:
lst.index(20)

0

In [104]:
lst.insert(0, 'chicken')
lst

['chicken', 20, 30, 40, 50, 60, 70, 80, 90, 100, 'hello']

In [105]:
lst.remove('hello')
print(lst)

['chicken', 20, 30, 40, 50, 60, 70, 80, 90, 100]


In [106]:
animals = ['ants', 'cats', 'dogs', 'badgers', 'elephants']
animals

['ants', 'cats', 'dogs', 'badgers', 'elephants']

In [107]:
animals.sort()
animals

['ants', 'badgers', 'cats', 'dogs', 'elephants']

In [108]:
animals.sort(reverse=True)
animals

['elephants', 'dogs', 'cats', 'badgers', 'ants']

In [109]:
animals = ['Ants', 'badgers', 'Cats', 'dogs', 'Elephants']
animals.sort(key=str.upper)
animals

['Ants', 'badgers', 'Cats', 'dogs', 'Elephants']

![](Capture24.png)

Mutable and Immutable Data Types
But lists and strings are different in an important way. A list value is a mutable data type: It can have values added, removed, or changed. However, a string is immutable: It cannot be changed. Trying to reassign a single character in a string results in a TypeError error, as you can see by entering the following into the interactive shell

In [110]:
name = 'Zophie a c  at'
name[7] = 'the'

TypeError: 'str' object does not support item assignment

The proper way to “mutate” a string is to use slicing and concatenation to build a new string by copying from parts of the old string. Enter the following into the interactive shell:


In [111]:
name = 'Zophie a cat'
newName = name[0:7] + 'the' + name[8:12]
print(name)
print(newName)

Zophie a cat
Zophie the cat


![](Capture16.png)

In [112]:
tup = ('hello', 42, 0.5)

In [113]:
tup[:1]

('hello',)

In [114]:
len(tup)

3

Tuples, like strings, are immutable. Tuples cannot have their values modified, appended, or removed. 

In [115]:
tup[0] = 0 # can not modify

TypeError: 'tuple' object does not support item assignment

In [116]:
type(tup)

tuple

![](Capture17.png)

In [117]:
myCat = {'size': 'fat', 'color': 'gray', 'disposition': 'loud'}

In [118]:
type(myCat)

dict

In [119]:
myCat['size']

'fat'

In [120]:
'My cat has ' + myCat['color'] + ' fur.'

'My cat has gray fur.'

In [121]:
myCat[0] 

KeyError: 0

In [122]:
eggs = {'name': 'Zophie', 'species': 'cat', 'age': '8'}
ham = {'species': 'cat', 'age': '8', 'name': 'Zophie'}
eggs == ham

True

In [123]:
for val in myCat.values():
    print(val)

fat
gray
loud


In [124]:
myCat.values()

dict_values(['fat', 'gray', 'loud'])

In [125]:
for key in myCat.keys():
    print(key)

size
color
disposition


In [126]:
myCat.keys()

dict_keys(['size', 'color', 'disposition'])

In [127]:
for i in myCat.items():
    print(i)

('size', 'fat')
('color', 'gray')
('disposition', 'loud')


In [128]:
myCat.items

<function dict.items>

In [129]:
myCat.items()

dict_items([('size', 'fat'), ('color', 'gray'), ('disposition', 'loud')])

In [130]:
'size' in myCat.keys()

True

In [131]:
'gray' not in myCat.values()

False

In [132]:
picnicItems = {'apples': 5, 'cups': 2}

In [133]:
'I am bringing ' + str(picnicItems.get('cups', 0)) + ' cups.'
#picnicItems.get('cups', 0) -> 0 is the default value only if the cups not in the dictionary

'I am bringing 2 cups.'

In [134]:
picnicItems.setdefault('colour', 'white')

'white'

In [135]:
picnicItems

{'apples': 5, 'cups': 2, 'colour': 'white'}

![](Capture18.png)

A Set is an unordered collection data type that is iterable, mutable, and has no duplicate elements.

In [136]:
normal_set = set(["a", "b","c"]) 

In [137]:
type(normal_set)

set

In [138]:
normal_set.add("d") 
  
print("Normal Set") 
print(normal_set) 

Normal Set
{'b', 'd', 'c', 'a'}


In [139]:
# A frozen set 
frozen_set = frozenset(["e", "f", "g"]) 
  
print("Frozen Set") 
print(frozen_set)

Frozen Set
frozenset({'e', 'g', 'f'})


In [140]:
type(frozen_set)

frozenset

In [141]:
frozen_set.add('d')

AttributeError: 'frozenset' object has no attribute 'add'

Frozen set is immutable data type

In [142]:
normal_set1 = set(["d", "b","c"]) 

In [143]:
#Union operation
normal_set | normal_set1
#uni = normal_set.union(normal_set1)

{'a', 'b', 'c', 'd'}

In [144]:
a = normal_set.intersection(normal_set1)
a

{'b', 'c', 'd'}

In [145]:
dif = normal_set.difference(normal_set1)
dif

{'a'}

In [146]:
dif = normal_set - normal_set1
dif

{'a'}

In [147]:
normal_set1.clear()
normal_set1

set()

![](Capture19.png)

Function is a set of statements that take inputs, do some specific computation and produces output

In [148]:
def evenodd( x ): 
    if (x % 2 == 0): 
        print("even")
    else: 
        print("odd")

In [149]:
evenodd(14) 
evenodd(33) 

even
odd


In Python every variable name is a reference. When we pass a variable to a function, a new reference to the object is created
Example are below

![](Capture20.png)

In [150]:
def f1(x): 
    x[0] = "change"

lst = [1, 1, 1, 3, 4, 15]  
f1(lst) 
print(lst)

['change', 1, 1, 3, 4, 15]


In [151]:
def f2(x): 
    # After below line link of x with previous object gets broken. A new object is assigned to x. 
    x = [20, 30, 40] 
    return(None) #it returns the whatever we give in the bracket to the called function


#Note that lst is not modified after function call. 
lst = [10, 11, 12, 13, 14, 15]  
f2(lst) #calling 
print(lst)  


[10, 11, 12, 13, 14, 15]


In [152]:
def myFun(x, y=50): 
    print("x: ", x) 
    print("y: ", y)
    print("id("+str(y)+") : ", id(y)) #id() returns the id of the object which is unique 
    #y = 50. the number 50 is an object not the variable y, y is the variable which is reference to the object "50"
    y = 0
    print("id(0) : ", id(y))

myFun

<function __main__.myFun(x, y=50)>

In [153]:
print(myFun(10,"good"))  #As you can see, None is printed in last line, which means that the myFun is returning nothing. So, it tells that the function "myFun" is returning nothing.

x:  10
y:  good
id(good) :  88508848
id(0) :  8791296864480
None


In [154]:
print(myFun(10,)) 

x:  10
y:  50
id(50) :  8791296866080
id(0) :  8791296864480
None


In [155]:
myFun(10,10)

x:  10
y:  10
id(10) :  8791296864800
id(0) :  8791296864480


In [156]:
def prompt(n): 
    value = int(input("Please enter integer "+ str(n) + ":")) 
    return value 
print("This program adds together two integers.") 
value1 = prompt(1) 
value2 = prompt(2)
sum = value1 + value2
print(value1, "+", value2, "=", sum)

This program adds together two integers.
Please enter integer 1:1
Please enter integer 2:2
1 + 2 = 3


In [157]:
def float1():
    x = 0.9 
    x += 0.1 
    if x == 1.0: 
        print("OK") 
    else: 
        print("NOT OK") 
a = float1()
b = float1()


OK
OK


In [158]:
print(id(a))
print(id(a))
print(id(float1()))

8791296392416
8791296392416
OK
8791296392416


In [159]:
def float2():
    i = 0.0 
    while i != 1.0: 
        print("i =", i) 
        i += 0.1 
float2()
#ctrl+c to stop the execution or click interrupt kernel

i = 0.0
i = 0.1
i = 0.2
i = 0.30000000000000004
i = 0.4
i = 0.5
i = 0.6
i = 0.7
i = 0.7999999999999999
i = 0.8999999999999999
i = 0.9999999999999999
i = 1.0999999999999999
i = 1.2
i = 1.3
i = 1.4000000000000001
i = 1.5000000000000002
i = 1.6000000000000003
i = 1.7000000000000004
i = 1.8000000000000005
i = 1.9000000000000006
i = 2.0000000000000004
i = 2.1000000000000005
i = 2.2000000000000006
i = 2.3000000000000007
i = 2.400000000000001
i = 2.500000000000001
i = 2.600000000000001
i = 2.700000000000001
i = 2.800000000000001
i = 2.9000000000000012
i = 3.0000000000000013
i = 3.1000000000000014
i = 3.2000000000000015
i = 3.3000000000000016
i = 3.4000000000000017
i = 3.5000000000000018
i = 3.600000000000002
i = 3.700000000000002
i = 3.800000000000002
i = 3.900000000000002
i = 4.000000000000002
i = 4.100000000000001
i = 4.200000000000001
i = 4.300000000000001
i = 4.4
i = 4.5
i = 4.6
i = 4.699999999999999
i = 4.799999999999999
i = 4.899999999999999
i = 4.999999999999998
i = 5.099999999999998
i

i = 119.89999999999746
i = 119.99999999999746
i = 120.09999999999745
i = 120.19999999999744
i = 120.29999999999744
i = 120.39999999999743
i = 120.49999999999743
i = 120.59999999999742
i = 120.69999999999742
i = 120.79999999999741
i = 120.8999999999974
i = 120.9999999999974
i = 121.0999999999974
i = 121.19999999999739
i = 121.29999999999738
i = 121.39999999999738
i = 121.49999999999737
i = 121.59999999999737
i = 121.69999999999736
i = 121.79999999999735
i = 121.89999999999735
i = 121.99999999999734
i = 122.09999999999734
i = 122.19999999999733
i = 122.29999999999733
i = 122.39999999999732
i = 122.49999999999731
i = 122.59999999999731
i = 122.6999999999973
i = 122.7999999999973
i = 122.89999999999729
i = 122.99999999999729
i = 123.09999999999728
i = 123.19999999999727
i = 123.29999999999727
i = 123.39999999999726
i = 123.49999999999726
i = 123.59999999999725
i = 123.69999999999725
i = 123.79999999999724
i = 123.89999999999723
i = 123.99999999999723
i = 124.09999999999722
i = 124.19999999

i = 230.99999999999116
i = 231.09999999999116
i = 231.19999999999115
i = 231.29999999999114
i = 231.39999999999114
i = 231.49999999999113
i = 231.59999999999113
i = 231.69999999999112
i = 231.79999999999112
i = 231.8999999999911
i = 231.9999999999911
i = 232.0999999999911
i = 232.1999999999911
i = 232.2999999999911
i = 232.39999999999108
i = 232.49999999999108
i = 232.59999999999107
i = 232.69999999999106
i = 232.79999999999106
i = 232.89999999999105
i = 232.99999999999105
i = 233.09999999999104
i = 233.19999999999104
i = 233.29999999999103
i = 233.39999999999102
i = 233.49999999999102
i = 233.599999999991
i = 233.699999999991
i = 233.799999999991
i = 233.899999999991
i = 233.999999999991
i = 234.09999999999098
i = 234.19999999999098
i = 234.29999999999097
i = 234.39999999999097
i = 234.49999999999096
i = 234.59999999999096
i = 234.69999999999095
i = 234.79999999999094
i = 234.89999999999094
i = 234.99999999999093
i = 235.09999999999093
i = 235.19999999999092
i = 235.29999999999092
i =

i = 344.9000000000099
i = 345.00000000000995
i = 345.10000000000997
i = 345.20000000001
i = 345.30000000001
i = 345.40000000001004
i = 345.50000000001006
i = 345.6000000000101
i = 345.7000000000101
i = 345.80000000001013
i = 345.90000000001015
i = 346.0000000000102
i = 346.1000000000102
i = 346.2000000000102
i = 346.30000000001024
i = 346.40000000001027
i = 346.5000000000103
i = 346.6000000000103
i = 346.70000000001033
i = 346.80000000001036
i = 346.9000000000104
i = 347.0000000000104
i = 347.1000000000104
i = 347.20000000001045
i = 347.30000000001047
i = 347.4000000000105
i = 347.5000000000105
i = 347.60000000001054
i = 347.70000000001056
i = 347.8000000000106
i = 347.9000000000106
i = 348.00000000001063
i = 348.10000000001065
i = 348.2000000000107
i = 348.3000000000107
i = 348.4000000000107
i = 348.50000000001074
i = 348.60000000001077
i = 348.7000000000108
i = 348.8000000000108
i = 348.90000000001083
i = 349.00000000001086
i = 349.1000000000109
i = 349.2000000000109
i = 349.30000000

i = 469.8000000000383
i = 469.90000000003835
i = 470.00000000003837
i = 470.1000000000384
i = 470.2000000000384
i = 470.30000000003844
i = 470.40000000003846
i = 470.5000000000385
i = 470.6000000000385
i = 470.70000000003853
i = 470.80000000003855
i = 470.9000000000386
i = 471.0000000000386
i = 471.1000000000386
i = 471.20000000003864
i = 471.30000000003866
i = 471.4000000000387
i = 471.5000000000387
i = 471.60000000003873
i = 471.70000000003876
i = 471.8000000000388
i = 471.9000000000388
i = 472.0000000000388
i = 472.10000000003885
i = 472.20000000003887
i = 472.3000000000389
i = 472.4000000000389
i = 472.50000000003894
i = 472.60000000003896
i = 472.700000000039
i = 472.800000000039
i = 472.90000000003903
i = 473.00000000003905
i = 473.1000000000391
i = 473.2000000000391
i = 473.3000000000391
i = 473.40000000003914
i = 473.50000000003917
i = 473.6000000000392
i = 473.7000000000392
i = 473.80000000003923
i = 473.90000000003926
i = 474.0000000000393
i = 474.1000000000393
i = 474.200000

i = 594.8000000000667
i = 594.9000000000667
i = 595.0000000000667
i = 595.1000000000668
i = 595.2000000000668
i = 595.3000000000668
i = 595.4000000000668
i = 595.5000000000668
i = 595.6000000000669
i = 595.7000000000669
i = 595.8000000000669
i = 595.9000000000669
i = 596.000000000067
i = 596.100000000067
i = 596.200000000067
i = 596.300000000067
i = 596.400000000067
i = 596.5000000000671
i = 596.6000000000671
i = 596.7000000000671
i = 596.8000000000671
i = 596.9000000000672
i = 597.0000000000672
i = 597.1000000000672
i = 597.2000000000672
i = 597.3000000000673
i = 597.4000000000673
i = 597.5000000000673
i = 597.6000000000673
i = 597.7000000000673
i = 597.8000000000674
i = 597.9000000000674
i = 598.0000000000674
i = 598.1000000000674
i = 598.2000000000675
i = 598.3000000000675
i = 598.4000000000675
i = 598.5000000000675
i = 598.6000000000676
i = 598.7000000000676
i = 598.8000000000676
i = 598.9000000000676
i = 599.0000000000676
i = 599.1000000000677
i = 599.2000000000677
i = 599.3000000

i = 719.8000000000951
i = 719.9000000000951
i = 720.0000000000952
i = 720.1000000000952
i = 720.2000000000952
i = 720.3000000000952
i = 720.4000000000952
i = 720.5000000000953
i = 720.6000000000953
i = 720.7000000000953
i = 720.8000000000953
i = 720.9000000000954
i = 721.0000000000954
i = 721.1000000000954
i = 721.2000000000954
i = 721.3000000000955
i = 721.4000000000955
i = 721.5000000000955
i = 721.6000000000955
i = 721.7000000000955
i = 721.8000000000956
i = 721.9000000000956
i = 722.0000000000956
i = 722.1000000000956
i = 722.2000000000957
i = 722.3000000000957
i = 722.4000000000957
i = 722.5000000000957
i = 722.6000000000957
i = 722.7000000000958
i = 722.8000000000958
i = 722.9000000000958
i = 723.0000000000958
i = 723.1000000000959
i = 723.2000000000959
i = 723.3000000000959
i = 723.4000000000959
i = 723.500000000096
i = 723.600000000096
i = 723.700000000096
i = 723.800000000096
i = 723.900000000096
i = 724.0000000000961
i = 724.1000000000961
i = 724.2000000000961
i = 724.3000000

i = 844.8000000001235
i = 844.9000000001236
i = 845.0000000001236
i = 845.1000000001236
i = 845.2000000001236
i = 845.3000000001236
i = 845.4000000001237
i = 845.5000000001237
i = 845.6000000001237
i = 845.7000000001237
i = 845.8000000001238
i = 845.9000000001238
i = 846.0000000001238
i = 846.1000000001238
i = 846.2000000001239
i = 846.3000000001239
i = 846.4000000001239
i = 846.5000000001239
i = 846.6000000001239
i = 846.700000000124
i = 846.800000000124
i = 846.900000000124
i = 847.000000000124
i = 847.100000000124
i = 847.2000000001241
i = 847.3000000001241
i = 847.4000000001241
i = 847.5000000001241
i = 847.6000000001242
i = 847.7000000001242
i = 847.8000000001242
i = 847.9000000001242
i = 848.0000000001243
i = 848.1000000001243
i = 848.2000000001243
i = 848.3000000001243
i = 848.4000000001244
i = 848.5000000001244
i = 848.6000000001244
i = 848.7000000001244
i = 848.8000000001244
i = 848.9000000001245
i = 849.0000000001245
i = 849.1000000001245
i = 849.2000000001245
i = 849.3000000

i = 969.7000000001519
i = 969.800000000152
i = 969.900000000152
i = 970.000000000152
i = 970.100000000152
i = 970.200000000152
i = 970.3000000001521
i = 970.4000000001521
i = 970.5000000001521
i = 970.6000000001521
i = 970.7000000001522
i = 970.8000000001522
i = 970.9000000001522
i = 971.0000000001522
i = 971.1000000001522
i = 971.2000000001523
i = 971.3000000001523
i = 971.4000000001523
i = 971.5000000001523
i = 971.6000000001524
i = 971.7000000001524
i = 971.8000000001524
i = 971.9000000001524
i = 972.0000000001525
i = 972.1000000001525
i = 972.2000000001525
i = 972.3000000001525
i = 972.4000000001525
i = 972.5000000001526
i = 972.6000000001526
i = 972.7000000001526
i = 972.8000000001526
i = 972.9000000001527
i = 973.0000000001527
i = 973.1000000001527
i = 973.2000000001527
i = 973.3000000001527
i = 973.4000000001528
i = 973.5000000001528
i = 973.6000000001528
i = 973.7000000001528
i = 973.8000000001529
i = 973.9000000001529
i = 974.0000000001529
i = 974.1000000001529
i = 974.2000000

i = 1094.7000000000999
i = 1094.8000000000998
i = 1094.9000000000997
i = 1095.0000000000996
i = 1095.1000000000995
i = 1095.2000000000994
i = 1095.3000000000993
i = 1095.4000000000992
i = 1095.5000000000991
i = 1095.600000000099
i = 1095.700000000099
i = 1095.8000000000989
i = 1095.9000000000988
i = 1096.0000000000987
i = 1096.1000000000986
i = 1096.2000000000985
i = 1096.3000000000984
i = 1096.4000000000983
i = 1096.5000000000982
i = 1096.6000000000981
i = 1096.700000000098
i = 1096.800000000098
i = 1096.9000000000979
i = 1097.0000000000978
i = 1097.1000000000977
i = 1097.2000000000976
i = 1097.3000000000975
i = 1097.4000000000974
i = 1097.5000000000973
i = 1097.6000000000972
i = 1097.7000000000971
i = 1097.800000000097
i = 1097.900000000097
i = 1098.0000000000969
i = 1098.1000000000968
i = 1098.2000000000967
i = 1098.3000000000966
i = 1098.4000000000965
i = 1098.5000000000964
i = 1098.6000000000963
i = 1098.7000000000962
i = 1098.8000000000961
i = 1098.900000000096
i = 1099.000000000

i = 1219.6999999999862
i = 1219.799999999986
i = 1219.899999999986
i = 1219.999999999986
i = 1220.0999999999858
i = 1220.1999999999857
i = 1220.2999999999856
i = 1220.3999999999855
i = 1220.4999999999854
i = 1220.5999999999854
i = 1220.6999999999853
i = 1220.7999999999852
i = 1220.899999999985
i = 1220.999999999985
i = 1221.099999999985
i = 1221.1999999999848
i = 1221.2999999999847
i = 1221.3999999999846
i = 1221.4999999999845
i = 1221.5999999999844
i = 1221.6999999999844
i = 1221.7999999999843
i = 1221.8999999999842
i = 1221.999999999984
i = 1222.099999999984
i = 1222.199999999984
i = 1222.2999999999838
i = 1222.3999999999837
i = 1222.4999999999836
i = 1222.5999999999835
i = 1222.6999999999834
i = 1222.7999999999834
i = 1222.8999999999833
i = 1222.9999999999832
i = 1223.099999999983
i = 1223.199999999983
i = 1223.299999999983
i = 1223.3999999999828
i = 1223.4999999999827
i = 1223.5999999999826
i = 1223.6999999999825
i = 1223.7999999999824
i = 1223.8999999999824
i = 1223.9999999999823


i = 1344.6999999998725
i = 1344.7999999998724
i = 1344.8999999998723
i = 1344.9999999998722
i = 1345.0999999998721
i = 1345.199999999872
i = 1345.299999999872
i = 1345.3999999998719
i = 1345.4999999998718
i = 1345.5999999998717
i = 1345.6999999998716
i = 1345.7999999998715
i = 1345.8999999998714
i = 1345.9999999998713
i = 1346.0999999998712
i = 1346.1999999998711
i = 1346.299999999871
i = 1346.399999999871
i = 1346.4999999998709
i = 1346.5999999998708
i = 1346.6999999998707
i = 1346.7999999998706
i = 1346.8999999998705
i = 1346.9999999998704
i = 1347.0999999998703
i = 1347.1999999998702
i = 1347.2999999998701
i = 1347.39999999987
i = 1347.49999999987
i = 1347.5999999998699
i = 1347.6999999998698
i = 1347.7999999998697
i = 1347.8999999998696
i = 1347.9999999998695
i = 1348.0999999998694
i = 1348.1999999998693
i = 1348.2999999998692
i = 1348.3999999998691
i = 1348.499999999869
i = 1348.599999999869
i = 1348.6999999998689
i = 1348.7999999998688
i = 1348.8999999998687
i = 1348.999999999868

i = 1469.599999999759
i = 1469.6999999997588
i = 1469.7999999997587
i = 1469.8999999997586
i = 1469.9999999997585
i = 1470.0999999997584
i = 1470.1999999997583
i = 1470.2999999997583
i = 1470.3999999997582
i = 1470.499999999758
i = 1470.599999999758
i = 1470.699999999758
i = 1470.7999999997578
i = 1470.8999999997577
i = 1470.9999999997576
i = 1471.0999999997575
i = 1471.1999999997574
i = 1471.2999999997573
i = 1471.3999999997573
i = 1471.4999999997572
i = 1471.599999999757
i = 1471.699999999757
i = 1471.799999999757
i = 1471.8999999997568
i = 1471.9999999997567
i = 1472.0999999997566
i = 1472.1999999997565
i = 1472.2999999997564
i = 1472.3999999997563
i = 1472.4999999997563
i = 1472.5999999997562
i = 1472.699999999756
i = 1472.799999999756
i = 1472.899999999756
i = 1472.9999999997558
i = 1473.0999999997557
i = 1473.1999999997556
i = 1473.2999999997555
i = 1473.3999999997554
i = 1473.4999999997553
i = 1473.5999999997553
i = 1473.6999999997552
i = 1473.799999999755
i = 1473.899999999755


i = 1594.5999999996452
i = 1594.6999999996451
i = 1594.799999999645
i = 1594.899999999645
i = 1594.9999999996448
i = 1595.0999999996448
i = 1595.1999999996447
i = 1595.2999999996446
i = 1595.3999999996445
i = 1595.4999999996444
i = 1595.5999999996443
i = 1595.6999999996442
i = 1595.7999999996441
i = 1595.899999999644
i = 1595.999999999644
i = 1596.0999999996438
i = 1596.1999999996438
i = 1596.2999999996437
i = 1596.3999999996436
i = 1596.4999999996435
i = 1596.5999999996434
i = 1596.6999999996433
i = 1596.7999999996432
i = 1596.8999999996431
i = 1596.999999999643
i = 1597.099999999643
i = 1597.1999999996428
i = 1597.2999999996428
i = 1597.3999999996427
i = 1597.4999999996426
i = 1597.5999999996425
i = 1597.6999999996424
i = 1597.7999999996423
i = 1597.8999999996422
i = 1597.9999999996421
i = 1598.099999999642
i = 1598.199999999642
i = 1598.2999999996418
i = 1598.3999999996418
i = 1598.4999999996417
i = 1598.5999999996416
i = 1598.6999999996415
i = 1598.7999999996414
i = 1598.8999999996

i = 1719.5999999995315
i = 1719.6999999995314
i = 1719.7999999995313
i = 1719.8999999995312
i = 1719.9999999995312
i = 1720.099999999531
i = 1720.199999999531
i = 1720.2999999995309
i = 1720.3999999995308
i = 1720.4999999995307
i = 1720.5999999995306
i = 1720.6999999995305
i = 1720.7999999995304
i = 1720.8999999995303
i = 1720.9999999995302
i = 1721.0999999995302
i = 1721.19999999953
i = 1721.29999999953
i = 1721.3999999995299
i = 1721.4999999995298
i = 1721.5999999995297
i = 1721.6999999995296
i = 1721.7999999995295
i = 1721.8999999995294
i = 1721.9999999995293
i = 1722.0999999995292
i = 1722.1999999995292
i = 1722.299999999529
i = 1722.399999999529
i = 1722.4999999995289
i = 1722.5999999995288
i = 1722.6999999995287
i = 1722.7999999995286
i = 1722.8999999995285
i = 1722.9999999995284
i = 1723.0999999995283
i = 1723.1999999995282
i = 1723.2999999995282
i = 1723.399999999528
i = 1723.499999999528
i = 1723.5999999995279
i = 1723.6999999995278
i = 1723.7999999995277
i = 1723.899999999527

i = 1844.5999999994178
i = 1844.6999999994177
i = 1844.7999999994177
i = 1844.8999999994176
i = 1844.9999999994175
i = 1845.0999999994174
i = 1845.1999999994173
i = 1845.2999999994172
i = 1845.399999999417
i = 1845.499999999417
i = 1845.599999999417
i = 1845.6999999994168
i = 1845.7999999994167
i = 1845.8999999994167
i = 1845.9999999994166
i = 1846.0999999994165
i = 1846.1999999994164
i = 1846.2999999994163
i = 1846.3999999994162
i = 1846.499999999416
i = 1846.599999999416
i = 1846.699999999416
i = 1846.7999999994158
i = 1846.8999999994157
i = 1846.9999999994156
i = 1847.0999999994156
i = 1847.1999999994155
i = 1847.2999999994154
i = 1847.3999999994153
i = 1847.4999999994152
i = 1847.599999999415
i = 1847.699999999415
i = 1847.799999999415
i = 1847.8999999994148
i = 1847.9999999994147
i = 1848.0999999994146
i = 1848.1999999994146
i = 1848.2999999994145
i = 1848.3999999994144
i = 1848.4999999994143
i = 1848.5999999994142
i = 1848.699999999414
i = 1848.799999999414
i = 1848.899999999414


i = 1969.4999999993042
i = 1969.5999999993041
i = 1969.699999999304
i = 1969.799999999304
i = 1969.8999999993039
i = 1969.9999999993038
i = 1970.0999999993037
i = 1970.1999999993036
i = 1970.2999999993035
i = 1970.3999999993034
i = 1970.4999999993033
i = 1970.5999999993032
i = 1970.6999999993031
i = 1970.799999999303
i = 1970.899999999303
i = 1970.9999999993029
i = 1971.0999999993028
i = 1971.1999999993027
i = 1971.2999999993026
i = 1971.3999999993025
i = 1971.4999999993024
i = 1971.5999999993023
i = 1971.6999999993022
i = 1971.7999999993021
i = 1971.899999999302
i = 1971.999999999302
i = 1972.0999999993019
i = 1972.1999999993018
i = 1972.2999999993017
i = 1972.3999999993016
i = 1972.4999999993015
i = 1972.5999999993014
i = 1972.6999999993013
i = 1972.7999999993012
i = 1972.8999999993011
i = 1972.999999999301
i = 1973.099999999301
i = 1973.1999999993009
i = 1973.2999999993008
i = 1973.3999999993007
i = 1973.4999999993006
i = 1973.5999999993005
i = 1973.6999999993004
i = 1973.7999999993

i = 2094.4999999991905
i = 2094.5999999991905
i = 2094.6999999991904
i = 2094.7999999991903
i = 2094.89999999919
i = 2094.99999999919
i = 2095.09999999919
i = 2095.19999999919
i = 2095.29999999919
i = 2095.3999999991897
i = 2095.4999999991896
i = 2095.5999999991895
i = 2095.6999999991895
i = 2095.7999999991894
i = 2095.8999999991893
i = 2095.999999999189
i = 2096.099999999189
i = 2096.199999999189
i = 2096.299999999189
i = 2096.399999999189
i = 2096.4999999991887
i = 2096.5999999991886
i = 2096.6999999991885
i = 2096.7999999991885
i = 2096.8999999991884
i = 2096.9999999991883
i = 2097.099999999188
i = 2097.199999999188
i = 2097.299999999188
i = 2097.399999999188
i = 2097.499999999188
i = 2097.5999999991877
i = 2097.6999999991876
i = 2097.7999999991875
i = 2097.8999999991875
i = 2097.9999999991874
i = 2098.0999999991873
i = 2098.199999999187
i = 2098.299999999187
i = 2098.399999999187
i = 2098.499999999187
i = 2098.599999999187
i = 2098.6999999991867
i = 2098.7999999991866
i = 2098.8999

i = 2219.499999999077
i = 2219.5999999990768
i = 2219.6999999990767
i = 2219.7999999990766
i = 2219.8999999990765
i = 2219.9999999990764
i = 2220.0999999990763
i = 2220.199999999076
i = 2220.299999999076
i = 2220.399999999076
i = 2220.499999999076
i = 2220.599999999076
i = 2220.6999999990758
i = 2220.7999999990757
i = 2220.8999999990756
i = 2220.9999999990755
i = 2221.0999999990754
i = 2221.1999999990753
i = 2221.299999999075
i = 2221.399999999075
i = 2221.499999999075
i = 2221.599999999075
i = 2221.699999999075
i = 2221.7999999990748
i = 2221.8999999990747
i = 2221.9999999990746
i = 2222.0999999990745
i = 2222.1999999990744
i = 2222.2999999990743
i = 2222.399999999074
i = 2222.499999999074
i = 2222.599999999074
i = 2222.699999999074
i = 2222.799999999074
i = 2222.8999999990738
i = 2222.9999999990737
i = 2223.0999999990736
i = 2223.1999999990735
i = 2223.2999999990734
i = 2223.3999999990733
i = 2223.499999999073
i = 2223.599999999073
i = 2223.699999999073
i = 2223.799999999073
i = 2223

i = 2344.499999998963
i = 2344.599999998963
i = 2344.699999998963
i = 2344.799999998963
i = 2344.899999998963
i = 2344.9999999989627
i = 2345.0999999989626
i = 2345.1999999989625
i = 2345.2999999989624
i = 2345.3999999989624
i = 2345.4999999989623
i = 2345.599999998962
i = 2345.699999998962
i = 2345.799999998962
i = 2345.899999998962
i = 2345.999999998962
i = 2346.0999999989617
i = 2346.1999999989616
i = 2346.2999999989615
i = 2346.3999999989614
i = 2346.4999999989614
i = 2346.5999999989613
i = 2346.699999998961
i = 2346.799999998961
i = 2346.899999998961
i = 2346.999999998961
i = 2347.099999998961
i = 2347.1999999989607
i = 2347.2999999989606
i = 2347.3999999989605
i = 2347.4999999989604
i = 2347.5999999989604
i = 2347.6999999989603
i = 2347.79999999896
i = 2347.89999999896
i = 2347.99999999896
i = 2348.09999999896
i = 2348.19999999896
i = 2348.2999999989597
i = 2348.3999999989596
i = 2348.4999999989595
i = 2348.5999999989594
i = 2348.6999999989594
i = 2348.7999999989593
i = 2348.8999

i = 2469.3999999988496
i = 2469.4999999988495
i = 2469.5999999988494
i = 2469.6999999988493
i = 2469.799999998849
i = 2469.899999998849
i = 2469.999999998849
i = 2470.099999998849
i = 2470.199999998849
i = 2470.2999999988488
i = 2470.3999999988487
i = 2470.4999999988486
i = 2470.5999999988485
i = 2470.6999999988484
i = 2470.7999999988483
i = 2470.899999998848
i = 2470.999999998848
i = 2471.099999998848
i = 2471.199999998848
i = 2471.299999998848
i = 2471.3999999988478
i = 2471.4999999988477
i = 2471.5999999988476
i = 2471.6999999988475
i = 2471.7999999988474
i = 2471.8999999988473
i = 2471.999999998847
i = 2472.099999998847
i = 2472.199999998847
i = 2472.299999998847
i = 2472.399999998847
i = 2472.4999999988468
i = 2472.5999999988467
i = 2472.6999999988466
i = 2472.7999999988465
i = 2472.8999999988464
i = 2472.9999999988463
i = 2473.099999998846
i = 2473.199999998846
i = 2473.299999998846
i = 2473.399999998846
i = 2473.499999998846
i = 2473.5999999988458
i = 2473.6999999988457
i = 2473

i = 2594.399999998736
i = 2594.499999998736
i = 2594.5999999987357
i = 2594.6999999987356
i = 2594.7999999987355
i = 2594.8999999987354
i = 2594.9999999987353
i = 2595.0999999987353
i = 2595.199999998735
i = 2595.299999998735
i = 2595.399999998735
i = 2595.499999998735
i = 2595.599999998735
i = 2595.6999999987347
i = 2595.7999999987346
i = 2595.8999999987345
i = 2595.9999999987344
i = 2596.0999999987343
i = 2596.1999999987343
i = 2596.299999998734
i = 2596.399999998734
i = 2596.499999998734
i = 2596.599999998734
i = 2596.699999998734
i = 2596.7999999987337
i = 2596.8999999987336
i = 2596.9999999987335
i = 2597.0999999987334
i = 2597.1999999987333
i = 2597.2999999987333
i = 2597.399999998733
i = 2597.499999998733
i = 2597.599999998733
i = 2597.699999998733
i = 2597.799999998733
i = 2597.8999999987327
i = 2597.9999999987326
i = 2598.0999999987325
i = 2598.1999999987324
i = 2598.2999999987323
i = 2598.3999999987323
i = 2598.499999998732
i = 2598.599999998732
i = 2598.699999998732
i = 2598

i = 2719.399999998622
i = 2719.499999998622
i = 2719.599999998622
i = 2719.699999998622
i = 2719.799999998622
i = 2719.8999999986218
i = 2719.9999999986217
i = 2720.0999999986216
i = 2720.1999999986215
i = 2720.2999999986214
i = 2720.3999999986213
i = 2720.499999998621
i = 2720.599999998621
i = 2720.699999998621
i = 2720.799999998621
i = 2720.899999998621
i = 2720.9999999986208
i = 2721.0999999986207
i = 2721.1999999986206
i = 2721.2999999986205
i = 2721.3999999986204
i = 2721.4999999986203
i = 2721.59999999862
i = 2721.69999999862
i = 2721.79999999862
i = 2721.89999999862
i = 2721.99999999862
i = 2722.0999999986198
i = 2722.1999999986197
i = 2722.2999999986196
i = 2722.3999999986195
i = 2722.4999999986194
i = 2722.5999999986193
i = 2722.699999998619
i = 2722.799999998619
i = 2722.899999998619
i = 2722.999999998619
i = 2723.099999998619
i = 2723.1999999986188
i = 2723.2999999986187
i = 2723.3999999986186
i = 2723.4999999986185
i = 2723.5999999986184
i = 2723.6999999986183
i = 2723.7999

i = 2844.3999999985085
i = 2844.4999999985084
i = 2844.5999999985083
i = 2844.6999999985082
i = 2844.799999998508
i = 2844.899999998508
i = 2844.999999998508
i = 2845.099999998508
i = 2845.199999998508
i = 2845.2999999985077
i = 2845.3999999985076
i = 2845.4999999985075
i = 2845.5999999985074
i = 2845.6999999985073
i = 2845.7999999985072
i = 2845.899999998507
i = 2845.999999998507
i = 2846.099999998507
i = 2846.199999998507
i = 2846.299999998507
i = 2846.3999999985067
i = 2846.4999999985066
i = 2846.5999999985065
i = 2846.6999999985064
i = 2846.7999999985063
i = 2846.8999999985062
i = 2846.999999998506
i = 2847.099999998506
i = 2847.199999998506
i = 2847.299999998506
i = 2847.399999998506
i = 2847.4999999985057
i = 2847.5999999985056
i = 2847.6999999985055
i = 2847.7999999985054
i = 2847.8999999985053
i = 2847.9999999985052
i = 2848.099999998505
i = 2848.199999998505
i = 2848.299999998505
i = 2848.399999998505
i = 2848.499999998505
i = 2848.5999999985047
i = 2848.6999999985046
i = 2848

i = 2969.299999998395
i = 2969.399999998395
i = 2969.4999999983947
i = 2969.5999999983947
i = 2969.6999999983946
i = 2969.7999999983945
i = 2969.8999999983944
i = 2969.9999999983943
i = 2970.099999998394
i = 2970.199999998394
i = 2970.299999998394
i = 2970.399999998394
i = 2970.499999998394
i = 2970.5999999983937
i = 2970.6999999983937
i = 2970.7999999983936
i = 2970.8999999983935
i = 2970.9999999983934
i = 2971.0999999983933
i = 2971.199999998393
i = 2971.299999998393
i = 2971.399999998393
i = 2971.499999998393
i = 2971.599999998393
i = 2971.6999999983927
i = 2971.7999999983927
i = 2971.8999999983926
i = 2971.9999999983925
i = 2972.0999999983924
i = 2972.1999999983923
i = 2972.299999998392
i = 2972.399999998392
i = 2972.499999998392
i = 2972.599999998392
i = 2972.699999998392
i = 2972.7999999983917
i = 2972.8999999983916
i = 2972.9999999983916
i = 2973.0999999983915
i = 2973.1999999983914
i = 2973.2999999983913
i = 2973.399999998391
i = 2973.499999998391
i = 2973.599999998391
i = 2973

i = 3094.2999999982812
i = 3094.399999998281
i = 3094.499999998281
i = 3094.599999998281
i = 3094.699999998281
i = 3094.799999998281
i = 3094.8999999982807
i = 3094.9999999982806
i = 3095.0999999982805
i = 3095.1999999982804
i = 3095.2999999982803
i = 3095.3999999982802
i = 3095.49999999828
i = 3095.59999999828
i = 3095.69999999828
i = 3095.79999999828
i = 3095.89999999828
i = 3095.9999999982797
i = 3096.0999999982796
i = 3096.1999999982795
i = 3096.2999999982794
i = 3096.3999999982793
i = 3096.4999999982792
i = 3096.599999998279
i = 3096.699999998279
i = 3096.799999998279
i = 3096.899999998279
i = 3096.999999998279
i = 3097.0999999982787
i = 3097.1999999982786
i = 3097.2999999982785
i = 3097.3999999982784
i = 3097.4999999982783
i = 3097.5999999982782
i = 3097.699999998278
i = 3097.799999998278
i = 3097.899999998278
i = 3097.999999998278
i = 3098.099999998278
i = 3098.1999999982777
i = 3098.2999999982776
i = 3098.3999999982775
i = 3098.4999999982774
i = 3098.5999999982773
i = 3098.6999

i = 3219.2999999981676
i = 3219.3999999981675
i = 3219.4999999981674
i = 3219.5999999981673
i = 3219.699999998167
i = 3219.799999998167
i = 3219.899999998167
i = 3219.999999998167
i = 3220.099999998167
i = 3220.1999999981667
i = 3220.2999999981666
i = 3220.3999999981665
i = 3220.4999999981665
i = 3220.5999999981664
i = 3220.6999999981663
i = 3220.799999998166
i = 3220.899999998166
i = 3220.999999998166
i = 3221.099999998166
i = 3221.199999998166
i = 3221.2999999981657
i = 3221.3999999981656
i = 3221.4999999981655
i = 3221.5999999981655
i = 3221.6999999981654
i = 3221.7999999981653
i = 3221.899999998165
i = 3221.999999998165
i = 3222.099999998165
i = 3222.199999998165
i = 3222.299999998165
i = 3222.3999999981647
i = 3222.4999999981646
i = 3222.5999999981645
i = 3222.6999999981645
i = 3222.7999999981644
i = 3222.8999999981643
i = 3222.999999998164
i = 3223.099999998164
i = 3223.199999998164
i = 3223.299999998164
i = 3223.399999998164
i = 3223.4999999981637
i = 3223.5999999981636
i = 3223

i = 3342.2999999980557
i = 3342.3999999980556
i = 3342.4999999980555
i = 3342.5999999980554
i = 3342.6999999980553
i = 3342.7999999980552
i = 3342.899999998055
i = 3342.999999998055
i = 3343.099999998055
i = 3343.199999998055
i = 3343.299999998055
i = 3343.3999999980547
i = 3343.4999999980546
i = 3343.5999999980545
i = 3343.6999999980544
i = 3343.7999999980543
i = 3343.899999998054
i = 3343.999999998054
i = 3344.099999998054
i = 3344.199999998054
i = 3344.299999998054
i = 3344.3999999980538
i = 3344.4999999980537
i = 3344.5999999980536
i = 3344.6999999980535
i = 3344.7999999980534
i = 3344.8999999980533
i = 3344.999999998053
i = 3345.099999998053
i = 3345.199999998053
i = 3345.299999998053
i = 3345.399999998053
i = 3345.4999999980528
i = 3345.5999999980527
i = 3345.6999999980526
i = 3345.7999999980525
i = 3345.8999999980524
i = 3345.9999999980523
i = 3346.099999998052
i = 3346.199999998052
i = 3346.299999998052
i = 3346.399999998052
i = 3346.499999998052
i = 3346.5999999980518
i = 3346

i = 3461.0999999979476
i = 3461.1999999979475
i = 3461.2999999979475
i = 3461.3999999979474
i = 3461.4999999979473
i = 3461.599999997947
i = 3461.699999997947
i = 3461.799999997947
i = 3461.899999997947
i = 3461.999999997947
i = 3462.0999999979467
i = 3462.1999999979466
i = 3462.2999999979465
i = 3462.3999999979465
i = 3462.4999999979464
i = 3462.5999999979463
i = 3462.699999997946
i = 3462.799999997946
i = 3462.899999997946
i = 3462.999999997946
i = 3463.099999997946
i = 3463.1999999979457
i = 3463.2999999979456
i = 3463.3999999979455
i = 3463.4999999979455
i = 3463.5999999979454
i = 3463.6999999979453
i = 3463.799999997945
i = 3463.899999997945
i = 3463.999999997945
i = 3464.099999997945
i = 3464.199999997945
i = 3464.2999999979447
i = 3464.3999999979446
i = 3464.4999999979445
i = 3464.5999999979445
i = 3464.6999999979444
i = 3464.7999999979443
i = 3464.899999997944
i = 3464.999999997944
i = 3465.099999997944
i = 3465.199999997944
i = 3465.299999997944
i = 3465.3999999979437
i = 3465

i = 3579.9999999978395
i = 3580.0999999978394
i = 3580.1999999978393
i = 3580.299999997839
i = 3580.399999997839
i = 3580.499999997839
i = 3580.599999997839
i = 3580.699999997839
i = 3580.7999999978388
i = 3580.8999999978387
i = 3580.9999999978386
i = 3581.0999999978385
i = 3581.1999999978384
i = 3581.2999999978383
i = 3581.399999997838
i = 3581.499999997838
i = 3581.599999997838
i = 3581.699999997838
i = 3581.799999997838
i = 3581.8999999978378
i = 3581.9999999978377
i = 3582.0999999978376
i = 3582.1999999978375
i = 3582.2999999978374
i = 3582.3999999978373
i = 3582.499999997837
i = 3582.599999997837
i = 3582.699999997837
i = 3582.799999997837
i = 3582.899999997837
i = 3582.9999999978368
i = 3583.0999999978367
i = 3583.1999999978366
i = 3583.2999999978365
i = 3583.3999999978364
i = 3583.4999999978363
i = 3583.599999997836
i = 3583.699999997836
i = 3583.799999997836
i = 3583.899999997836
i = 3583.999999997836
i = 3584.0999999978358
i = 3584.1999999978357
i = 3584.2999999978356
i = 3584

i = 3699.9999999977304
i = 3700.0999999977303
i = 3700.19999999773
i = 3700.29999999773
i = 3700.39999999773
i = 3700.49999999773
i = 3700.59999999773
i = 3700.6999999977297
i = 3700.7999999977296
i = 3700.8999999977295
i = 3700.9999999977294
i = 3701.0999999977294
i = 3701.1999999977293
i = 3701.299999997729
i = 3701.399999997729
i = 3701.499999997729
i = 3701.599999997729
i = 3701.699999997729
i = 3701.7999999977287
i = 3701.8999999977286
i = 3701.9999999977285
i = 3702.0999999977284
i = 3702.1999999977284
i = 3702.2999999977283
i = 3702.399999997728
i = 3702.499999997728
i = 3702.599999997728
i = 3702.699999997728
i = 3702.799999997728
i = 3702.8999999977277
i = 3702.9999999977276
i = 3703.0999999977275
i = 3703.1999999977274
i = 3703.2999999977274
i = 3703.3999999977273
i = 3703.499999997727
i = 3703.599999997727
i = 3703.699999997727
i = 3703.799999997727
i = 3703.899999997727
i = 3703.9999999977267
i = 3704.0999999977266
i = 3704.1999999977265
i = 3704.2999999977264
i = 3704.3999

i = 3823.0999999976184
i = 3823.1999999976183
i = 3823.299999997618
i = 3823.399999997618
i = 3823.499999997618
i = 3823.599999997618
i = 3823.699999997618
i = 3823.7999999976178
i = 3823.8999999976177
i = 3823.9999999976176
i = 3824.0999999976175
i = 3824.1999999976174
i = 3824.2999999976173
i = 3824.399999997617
i = 3824.499999997617
i = 3824.599999997617
i = 3824.699999997617
i = 3824.799999997617
i = 3824.8999999976168
i = 3824.9999999976167
i = 3825.0999999976166
i = 3825.1999999976165
i = 3825.2999999976164
i = 3825.3999999976163
i = 3825.499999997616
i = 3825.599999997616
i = 3825.699999997616
i = 3825.799999997616
i = 3825.899999997616
i = 3825.9999999976158
i = 3826.0999999976157
i = 3826.1999999976156
i = 3826.2999999976155
i = 3826.3999999976154
i = 3826.4999999976153
i = 3826.599999997615
i = 3826.699999997615
i = 3826.799999997615
i = 3826.899999997615
i = 3826.999999997615
i = 3827.0999999976148
i = 3827.1999999976147
i = 3827.2999999976146
i = 3827.3999999976145
i = 3827

i = 3944.0999999975083
i = 3944.1999999975083
i = 3944.299999997508
i = 3944.399999997508
i = 3944.499999997508
i = 3944.599999997508
i = 3944.699999997508
i = 3944.7999999975077
i = 3944.8999999975076
i = 3944.9999999975075
i = 3945.0999999975074
i = 3945.1999999975073
i = 3945.2999999975073
i = 3945.399999997507
i = 3945.499999997507
i = 3945.599999997507
i = 3945.699999997507
i = 3945.799999997507
i = 3945.8999999975067
i = 3945.9999999975066
i = 3946.0999999975065
i = 3946.1999999975064
i = 3946.2999999975063
i = 3946.3999999975063
i = 3946.499999997506
i = 3946.599999997506
i = 3946.699999997506
i = 3946.799999997506
i = 3946.899999997506
i = 3946.9999999975057
i = 3947.0999999975056
i = 3947.1999999975055
i = 3947.2999999975054
i = 3947.3999999975053
i = 3947.4999999975053
i = 3947.599999997505
i = 3947.699999997505
i = 3947.799999997505
i = 3947.899999997505
i = 3947.999999997505
i = 3948.0999999975047
i = 3948.1999999975046
i = 3948.2999999975045
i = 3948.3999999975044
i = 3948

i = 4066.299999997397
i = 4066.399999997397
i = 4066.499999997397
i = 4066.599999997397
i = 4066.699999997397
i = 4066.7999999973968
i = 4066.8999999973967
i = 4066.9999999973966
i = 4067.0999999973965
i = 4067.1999999973964
i = 4067.2999999973963
i = 4067.399999997396
i = 4067.499999997396
i = 4067.599999997396
i = 4067.699999997396
i = 4067.799999997396
i = 4067.8999999973958
i = 4067.9999999973957
i = 4068.0999999973956
i = 4068.1999999973955
i = 4068.2999999973954
i = 4068.3999999973953
i = 4068.499999997395
i = 4068.599999997395
i = 4068.699999997395
i = 4068.799999997395
i = 4068.899999997395
i = 4068.9999999973948
i = 4069.0999999973947
i = 4069.1999999973946
i = 4069.2999999973945
i = 4069.3999999973944
i = 4069.4999999973943
i = 4069.599999997394
i = 4069.699999997394
i = 4069.799999997394
i = 4069.899999997394
i = 4069.999999997394
i = 4070.0999999973938
i = 4070.1999999973937
i = 4070.2999999973936
i = 4070.3999999973935
i = 4070.4999999973934
i = 4070.5999999973933
i = 4070

i = 4185.999999997697
i = 4186.0999999976975
i = 4186.199999997698
i = 4186.299999997698
i = 4186.399999997699
i = 4186.499999997699
i = 4186.599999997699
i = 4186.6999999977
i = 4186.7999999977
i = 4186.8999999977
i = 4186.999999997701
i = 4187.099999997701
i = 4187.1999999977015
i = 4187.299999997702
i = 4187.399999997702
i = 4187.499999997703
i = 4187.599999997703
i = 4187.699999997703
i = 4187.799999997704
i = 4187.899999997704
i = 4187.999999997704
i = 4188.099999997705
i = 4188.199999997705
i = 4188.2999999977055
i = 4188.399999997706
i = 4188.499999997706
i = 4188.599999997707
i = 4188.699999997707
i = 4188.799999997707
i = 4188.899999997708
i = 4188.999999997708
i = 4189.099999997708
i = 4189.199999997709
i = 4189.299999997709
i = 4189.3999999977095
i = 4189.49999999771
i = 4189.59999999771
i = 4189.699999997711
i = 4189.799999997711
i = 4189.899999997711
i = 4189.999999997712
i = 4190.099999997712
i = 4190.199999997712
i = 4190.299999997713
i = 4190.399999997713
i = 4190.49999

i = 4302.999999998123
i = 4303.099999998123
i = 4303.1999999981235
i = 4303.299999998124
i = 4303.399999998124
i = 4303.499999998125
i = 4303.599999998125
i = 4303.699999998125
i = 4303.799999998126
i = 4303.899999998126
i = 4303.999999998126
i = 4304.099999998127
i = 4304.199999998127
i = 4304.2999999981275
i = 4304.399999998128
i = 4304.499999998128
i = 4304.599999998129
i = 4304.699999998129
i = 4304.799999998129
i = 4304.89999999813
i = 4304.99999999813
i = 4305.09999999813
i = 4305.199999998131
i = 4305.299999998131
i = 4305.3999999981315
i = 4305.499999998132
i = 4305.599999998132
i = 4305.699999998133
i = 4305.799999998133
i = 4305.899999998133
i = 4305.999999998134
i = 4306.099999998134
i = 4306.199999998134
i = 4306.299999998135
i = 4306.399999998135
i = 4306.4999999981355
i = 4306.599999998136
i = 4306.699999998136
i = 4306.799999998137
i = 4306.899999998137
i = 4306.999999998137
i = 4307.099999998138
i = 4307.199999998138
i = 4307.299999998138
i = 4307.399999998139
i = 4307.

i = 4423.19999999856
i = 4423.29999999856
i = 4423.399999998561
i = 4423.499999998561
i = 4423.5999999985615
i = 4423.699999998562
i = 4423.799999998562
i = 4423.899999998563
i = 4423.999999998563
i = 4424.099999998563
i = 4424.199999998564
i = 4424.299999998564
i = 4424.399999998564
i = 4424.499999998565
i = 4424.599999998565
i = 4424.6999999985655
i = 4424.799999998566
i = 4424.899999998566
i = 4424.999999998567
i = 4425.099999998567
i = 4425.199999998567
i = 4425.299999998568
i = 4425.399999998568
i = 4425.4999999985685
i = 4425.599999998569
i = 4425.699999998569
i = 4425.79999999857
i = 4425.89999999857
i = 4425.99999999857
i = 4426.099999998571
i = 4426.199999998571
i = 4426.299999998571
i = 4426.399999998572
i = 4426.499999998572
i = 4426.5999999985725
i = 4426.699999998573
i = 4426.799999998573
i = 4426.899999998574
i = 4426.999999998574
i = 4427.099999998574
i = 4427.199999998575
i = 4427.299999998575
i = 4427.399999998575
i = 4427.499999998576
i = 4427.599999998576
i = 4427.69

i = 4544.199999999
i = 4544.299999999001
i = 4544.399999999001
i = 4544.499999999001
i = 4544.599999999002
i = 4544.699999999002
i = 4544.7999999990025
i = 4544.899999999003
i = 4544.999999999003
i = 4545.099999999004
i = 4545.199999999004
i = 4545.299999999004
i = 4545.399999999005
i = 4545.499999999005
i = 4545.599999999005
i = 4545.699999999006
i = 4545.799999999006
i = 4545.8999999990065
i = 4545.999999999007
i = 4546.099999999007
i = 4546.199999999008
i = 4546.299999999008
i = 4546.399999999008
i = 4546.499999999009
i = 4546.599999999009
i = 4546.699999999009
i = 4546.79999999901
i = 4546.89999999901
i = 4546.9999999990105
i = 4547.099999999011
i = 4547.199999999011
i = 4547.299999999012
i = 4547.399999999012
i = 4547.499999999012
i = 4547.599999999013
i = 4547.699999999013
i = 4547.799999999013
i = 4547.899999999014
i = 4547.999999999014
i = 4548.0999999990145
i = 4548.199999999015
i = 4548.299999999015
i = 4548.399999999016
i = 4548.499999999016
i = 4548.599999999016
i = 4548.69

i = 4668.299999999452
i = 4668.399999999452
i = 4668.4999999994525
i = 4668.599999999453
i = 4668.699999999453
i = 4668.799999999454
i = 4668.899999999454
i = 4668.999999999454
i = 4669.099999999455
i = 4669.199999999455
i = 4669.299999999455
i = 4669.399999999456
i = 4669.499999999456
i = 4669.5999999994565
i = 4669.699999999457
i = 4669.799999999457
i = 4669.899999999458
i = 4669.999999999458
i = 4670.099999999458
i = 4670.199999999459
i = 4670.299999999459
i = 4670.399999999459
i = 4670.49999999946
i = 4670.59999999946
i = 4670.6999999994605
i = 4670.799999999461
i = 4670.899999999461
i = 4670.999999999462
i = 4671.099999999462
i = 4671.199999999462
i = 4671.299999999463
i = 4671.399999999463
i = 4671.499999999463
i = 4671.599999999464
i = 4671.699999999464
i = 4671.7999999994645
i = 4671.899999999465
i = 4671.999999999465
i = 4672.099999999466
i = 4672.199999999466
i = 4672.299999999466
i = 4672.399999999467
i = 4672.499999999467
i = 4672.599999999467
i = 4672.699999999468
i = 4672

i = 4788.299999999888
i = 4788.399999999889
i = 4788.499999999889
i = 4788.599999999889
i = 4788.69999999989
i = 4788.79999999989
i = 4788.8999999998905
i = 4788.999999999891
i = 4789.099999999891
i = 4789.199999999892
i = 4789.299999999892
i = 4789.399999999892
i = 4789.499999999893
i = 4789.599999999893
i = 4789.699999999893
i = 4789.799999999894
i = 4789.899999999894
i = 4789.9999999998945
i = 4790.099999999895
i = 4790.199999999895
i = 4790.299999999896
i = 4790.399999999896
i = 4790.499999999896
i = 4790.599999999897
i = 4790.699999999897
i = 4790.799999999897
i = 4790.899999999898
i = 4790.999999999898
i = 4791.0999999998985
i = 4791.199999999899
i = 4791.299999999899
i = 4791.3999999999
i = 4791.4999999999
i = 4791.5999999999
i = 4791.699999999901
i = 4791.799999999901
i = 4791.899999999901
i = 4791.999999999902
i = 4792.099999999902
i = 4792.1999999999025
i = 4792.299999999903
i = 4792.399999999903
i = 4792.499999999904
i = 4792.599999999904
i = 4792.699999999904
i = 4792.79999

KeyboardInterrupt: 

![](Capture21.png)

Better not to use '==' and '!=' in floating point conditions.

USe '<', '>' to avoid the infinite loop

In [160]:
def f(iman=[]):
    
    iman.append(1)
    print("In function -",id(iman), "- ", iman)
 
    return iman
   
print(id(iman), "- ", iman)
print()

f() # 1st calling
print("out function -",id(iman), "- ", iman)
print()


f() # 2nd calling
print("out function -",id(iman), "- ", iman)
print()

f(iman=[2]) # 3rd calling
print("out function -",id(iman), "- ", iman)
print()


f() # 4th calling
print("out function -",id(iman), "- ", iman)

NameError: name 'iman' is not defined

In [161]:
This is because, the default parameter is the list, which is mutable(changing) datatype.

And the iman variable inside and outside the function f() are not same.
i.e., The iman in the function f() is different from the iman which is in outside of the function f()

Consider the below example -

SyntaxError: invalid syntax (<ipython-input-161-4a1e46c0ea8c>, line 1)

In [162]:
A= 10

def funct():
    print("In function")
    A = 20 
    print(A)

funct()
print(A)

In function
20
10


The variable A present inside and outside the function are not same.

Because, the variable A present inside the funct() is local variable and its scope only within the function

And, the variable A outside the funct() is like a global variable but its scope only in outised of the function funct()

In [163]:
def some_func(): 
    print("Inside some_func") 
    inner_var = 10
    print(inner_var)
some_func() 
print(inner_var)
# This make error because the inner_var is defined in the function "some_func" and, we accessing the inner_var from outside. This is not possible.

Inside some_func
10


NameError: name 'inner_var' is not defined

In [164]:
def some_func(): 
    print("Inside some_func") 
    inner_var = 10
    print(inner_var)
    
def some_func1():
    print("Inside some_func1") 
    print(inner_var) # Accessing other function variables in this function makes error
    
some_func() 
some_func1()


Inside some_func
10
Inside some_func1


NameError: name 'inner_var' is not defined

In [165]:
def some_func(): 
    print("Inside some_func") 
    print(outer_var)

outer_var = 10
some_func() 

Inside some_func
10


###### Global Variable

In [166]:
def some_func(): 
    global inner_var #If we want a varaible as global, then we need to define that variable as global in begining of the function
    print("Inside some_func") 
    inner_var = 10
    print(inner_var)

inner_var =5
some_func() 
print("Outside some_func") 
print(inner_var)

#If a variable can be reached anywhere in the code is called a global variable. If a variable is known only inside the scope, we call it a local variable.


Inside some_func
10
Outside some_func
10


![](Capture22.png)

In [167]:
def square(y): 
    return y*y; 

print(square(5)) 

25


In [168]:
#lambda arguments: expression
g = lambda x: x*x 
print(g(5))
#This function can have any number of arguments but only one expression, which is evaluated and returned

25


In [169]:
li = [5, 7, 22, 97, 54, 62, 77, 23, 73, 61] 
final_list = list(filter(lambda x: (x%2 != 0) , li)) 
print(final_list) 

[5, 7, 97, 77, 23, 73, 61]


In [170]:
li = [5, 7, 22, 97, 54, 62, 77, 23, 73, 61] 
final_list = list(map(lambda x: x*2 , li)) 
print(final_list) 


[10, 14, 44, 194, 108, 124, 154, 46, 146, 122]


In [171]:
li = [5, 7, 22, 97, 54, 62, 77, 23, 73, 61] 
final_list = list(filter(lambda x: x*2 , li)) 
print(final_list) 
#It is not a syntax error but logic error. Because, the filter only filters with the given condition. 
#Here, we gave a statement(x*2) not a condition.
#That is why, it print the list as it is.

[5, 7, 22, 97, 54, 62, 77, 23, 73, 61]


![](Capture23.png)

In [172]:
def myFun(*argv):  # argv can handle more number of arguments
    for arg in argv:  
        print (arg) 
    
myFun('Hello,', 'Have ', 'a ', 'good ', 'day') 

Hello,
Have 
a 
good 
day


In [173]:
def myFun(**kwargs):  
    for key, value in kwargs.items(): 
        print (str(key) + " - " + str(value)) 
  
# Driver code 
myFun(name= 'kwargs', habbit ='pass arguments as a block or like a dictonary', synatx=' any variable prefixed with double * (**)')  

name - kwargs
habbit - pass arguments as a block or like a dictonary
synatx -  any variable prefixed with double * (**)


![](Capture25.png)

In [174]:
"""
try:
    
   You do your operations here; If it throws any error, then the except block below catch and, handle to execute the code without interept
    
except:
   If there is any exception, then execute this block.
   
"""

'\ntry:\n    \n   You do your operations here; If it throws any error, then the except block below catch and, handle to execute the code without interept\n    \nexcept:\n   If there is any exception, then execute this block.\n   \n'

In [175]:
try:
    a = 10/0
    print(a)
    
except Exception as ex:
    print(type(ex).__name__, "-", ex.args)


ZeroDivisionError - ('division by zero',)


In [176]:
try:
    a = 10/0
    print(a)
    
except Exception as ex:
    print(Exception)
    print(type(ex).__name__, "-", ex.args, "-",)

finally:
    print("Final block, if needed") # The finally will execute no mater the try block throws error(s) or not

<class 'Exception'>
ZeroDivisionError - ('division by zero',) -
Final block, if needed


In [177]:
try:
    a = 10/10
    print(a)
    
except Exception as ex:
    print(type(ex).__name__, "-", ex.args)

else:
    print("Final block, if needed") # The else block will execute only if the try block does not throws any error(s)

1.0
Final block, if needed


In [178]:
"""
The list of Exception in python:
    
Exception	Descripton
BaseException	This is the root exception for all others
GeneratorExit	Raised by close() method of generators for terminating iteration
KeyboardInterrupt	Raised by the interrupt key
SystemExit	Program exit
Exception	Root for all non-exiting exceptions
StopIteration	Raised to stop an iteration action
StandardError	Base class for all built-in exceptions
ArithmeticError	Base for all arithmetic exceptions
FloatingPointError	Raised when a floating-point operation fails
OverflowError	Arithmetic operations that are too large
ZeroDivisoinError	Division or modulo operation with zero is divisor
AssertionError	Used when an assert statement fails
AttributeError	Attribute reference or failure to assign correctly
EnvironmentError	An error occurred outside of Python
IOError	Error in Input/Output operation
OSError	An error occurred in the os module
EOFError	input() or raw_input() tried to read past the end of a file
ImportError	Import failed to find module or name
LookupError	Base class for IndexError and KeyError
IndexError	A sequence index goes out of range
KeyError	Referenced a non-existent mapping (dict) key
MemoryError	Memory exhausted
NameError	Failure to find a local or global name
UnboundLocalError	Unassigned local variable is referenced
ReferenceError	Attempt to access a garbage-collected object
RuntimeError	Obsolete catch-all error
NotImplementedError	Raised when a feature is not implemented
SyntaxError	Parser encountered a syntax error
IndentationError	Parser encountered an indentation issue
TabError	Incorrect mixture of tabs and spaces
SystemError	Non-fatal interpreter error
TypeError	Inappropriate type was passed to a built-in operator or function
ValueError	Argument error not covered by TypeError or a more precise error
Warning	Base for all warnings
"""

'\nThe list of Exception in python:\n    \nException\tDescripton\nBaseException\tThis is the root exception for all others\nGeneratorExit\tRaised by close() method of generators for terminating iteration\nKeyboardInterrupt\tRaised by the interrupt key\nSystemExit\tProgram exit\nException\tRoot for all non-exiting exceptions\nStopIteration\tRaised to stop an iteration action\nStandardError\tBase class for all built-in exceptions\nArithmeticError\tBase for all arithmetic exceptions\nFloatingPointError\tRaised when a floating-point operation fails\nOverflowError\tArithmetic operations that are too large\nZeroDivisoinError\tDivision or modulo operation with zero is divisor\nAssertionError\tUsed when an assert statement fails\nAttributeError\tAttribute reference or failure to assign correctly\nEnvironmentError\tAn error occurred outside of Python\nIOError\tError in Input/Output operation\nOSError\tAn error occurred in the os module\nEOFError\tinput() or raw_input() tried to read past the en

![](Capture26.png)

In [179]:
import l # Reffer the l.py file in this directory

In [180]:
dir(l)

['__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'add',
 'mod']

In [181]:
l.__builtins__ #prints the builtin functions in python which are access by this function l

{'__name__': 'builtins',
 '__doc__': "Built-in functions, exceptions, and other objects.\n\nNoteworthy: None is the `nil' object; Ellipsis represents `...' in slices.",
 '__package__': '',
 '__loader__': _frozen_importlib.BuiltinImporter,
 '__spec__': ModuleSpec(name='builtins', loader=<class '_frozen_importlib.BuiltinImporter'>),
 '__build_class__': <function __build_class__>,
 '__import__': <function __import__>,
 'abs': <function abs(x, /)>,
 'all': <function all(iterable, /)>,
 'any': <function any(iterable, /)>,
 'ascii': <function ascii(obj, /)>,
 'bin': <function bin(number, /)>,
 'breakpoint': <function breakpoint>,
 'callable': <function callable(obj, /)>,
 'chr': <function chr(i, /)>,
 'compile': <function compile(source, filename, mode, flags=0, dont_inherit=False, optimize=-1)>,
 'delattr': <function delattr(obj, name, /)>,
 'dir': <function dir>,
 'divmod': <function divmod(x, y, /)>,
 'eval': <function eval(source, globals=None, locals=None, /)>,
 'exec': <function exec(s

In [182]:
l.__cached__ 

'C:\\Users\\softloft\\images\\__pycache__\\l.cpython-37.pyc'

# cached file. The l.py is compiled to .pyc file formate which means the python code is converted to byte code. The byte code helps to run faster when executing. 

In [183]:
import dis # this module shows you how the python code(l.py) is stored as byte code(l.cpython-37.pyc) .
dis.dis(l)

Disassembly of add:
  9           0 LOAD_FAST                0 (x)
              2 LOAD_FAST                1 (y)
              4 BINARY_ADD
              6 RETURN_VALUE

Disassembly of mod:
  6           0 LOAD_GLOBAL              0 (print)
              2 LOAD_CONST               1 ('hello')
              4 CALL_FUNCTION            1
              6 POP_TOP
              8 LOAD_CONST               0 (None)
             10 RETURN_VALUE



In [184]:
l.__doc__ # prints the content in the Triple-double quote

'\nThis module helps you to learn the modules in python.\n'

In [185]:
l.__file__

'C:\\Users\\softloft\\images\\l.py'

In [186]:
l.__loader__ # loader is an attribute which returns the object. The object represnts the module l is loaded in the memory.
#check this link for what is loader https://stackoverflow.com/questions/22185888/pythons-loader-what-is-it

In [187]:
l.__name__

'l'

In [188]:
l.__package__

''

In [189]:
l.__spec__

ModuleSpec(name='l', loader=<_frozen_importlib_external.SourceFileLoader object at 0x0000000005453408>, origin='C:\\Users\\softloft\\images\\l.py')

In [190]:
l.mod()

hello


In [191]:
l.add(2,2)

4

When import statement is executed, the python interepter looks the package (in this case, the package is l) in the following places

1. The current directory.

2. If the module isn't found, Python then searches each directory in the shell variable PYTHONPATH.

3. If all else fails, Python checks the default path. On UNIX, this default path is normally /usr/local/lib/python/.



In [192]:
import pack

In [193]:
dir(pack) # In this the module1 and, module2 file is not shown, we need to import module1 and, module2 explicitly

['__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__']

In [194]:
print(pack.__builtins__)
print(pack.__cached__)
print(pack.__doc__)
print(pack.__file__)
print(pack.__loader__)
print(pack.__name__)
print(pack.__package__)
print(pack.__path__)
print(pack.__spec__)

{'__name__': 'builtins', '__doc__': "Built-in functions, exceptions, and other objects.\n\nNoteworthy: None is the `nil' object; Ellipsis represents `...' in slices.", '__package__': '', '__loader__': <class '_frozen_importlib.BuiltinImporter'>, '__spec__': ModuleSpec(name='builtins', loader=<class '_frozen_importlib.BuiltinImporter'>), '__build_class__': <built-in function __build_class__>, '__import__': <built-in function __import__>, 'abs': <built-in function abs>, 'all': <built-in function all>, 'any': <built-in function any>, 'ascii': <built-in function ascii>, 'bin': <built-in function bin>, 'breakpoint': <built-in function breakpoint>, 'callable': <built-in function callable>, 'chr': <built-in function chr>, 'compile': <built-in function compile>, 'delattr': <built-in function delattr>, 'dir': <built-in function dir>, 'divmod': <built-in function divmod>, 'eval': <built-in function eval>, 'exec': <built-in function exec>, 'format': <built-in function format>, 'getattr': <built-i

In [195]:
help(pack)

Help on package pack:

NAME
    pack - PAckage pack

PACKAGE CONTENTS
    module1
    module2

FILE
    c:\users\softloft\images\pack\__init__.py




In [196]:
from pack import module1

In [197]:
import pack.module2 as mod2

In [198]:
help(pack)

Help on package pack:

NAME
    pack - PAckage pack

PACKAGE CONTENTS
    module1
    module2

FILE
    c:\users\softloft\images\pack\__init__.py




In [199]:
dir(pack)

['__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 'module1',
 'module2']

In [200]:
pack.module1.hello()

hello


In [201]:
module1.hello()

hello


In [202]:
pack.module2.greetings()

greetings


In [203]:
mod2.greetings()

greetings


In [204]:
mod2.mul(2,2)

4

![](Capture27.png)

In [205]:
class car(): 
    
    val = 20
    def show(model, color): 
        print("Model is", model ) 
        print("color is", color ) 
          
car.show("audi a4", "blue") 
car.show("ferrari 488", "green")
print(car.val)


Model is audi a4
color is blue
Model is ferrari 488
color is green
20


In [206]:
class car(): 
    
    val = 20
    @classmethod
    def show(cls, model, color): 
        print("Model is", model ) 
        print("color is", color ) 
        print(cls.val) # we can access the class variables inside the class methods by defining @classmethod
        # print(car.val) prints 20 
car.show("audi a4", "blue") 
car.show("ferrari 488", "green")


Model is audi a4
color is blue
20
Model is ferrari 488
color is green
20


In [207]:
class car(): 
                
    def show(model, color): 
        print("Model is", model ) 
        print("color is", color ) 
          
# Creating instance to the calss car
a = car() # a is the instance to the class car
a.show("bmw x4", "red")

TypeError: show() takes 2 positional arguments but 3 were given

In [208]:
The reason for the error -

When a.show("bmw x4", "red") is executed,

Python convert it into car.show(a, "bmw x4", "red") # class_name.method_name(instance, argument1, argument2, ...)

Now, the call 'car.show(a, "bmw x4", "red")' has three arguments such as,
1. a - instance of the class
2. "bmw x4" - model 
3. "red" - color


But, in defining the show() method in class car, we just gave only two argument(model and, color) - "def show(model, color):"

To resolve this error, we add one more argument in the method definition to represent the instance. Like -"def show(self, model, color):" 

Now, the call 'car.show(a, "d", "e")' has three arguments and, also the definition 'def show(self, model, color):' has three arguments

Number of arguments matched and, works fine

SyntaxError: invalid syntax (<ipython-input-208-0921c53347e2>, line 1)

In [209]:
class car(): 
      
    # init method which is the constructor 
    def __init__(self, model, color): 
        self.model = model 
        self.color = color 
          
    def show(self): 
        print("Model is", self.model ) 
        print("color is", self.color )
        val =10

# audi and, ferrari are the Instances of the class car

audi = car("audi a4", "blue")  
# this call is converted to car.__init__(audi, "audi a4", "blue")
# Because of this convertion, we need to define the __init__ method with three arguments which is (self, model, color).
#In the place of self, the instance will be replaced

ferrari = car("ferrari 488", "green") 
# For this instance,
# car.__init__(ferrari, "ferrari 488", "green")


print("object audi")
audi.show()     
# car.show(audi)
# In this call the audi instance is passed as argument, this is why we defined the show method in class car with one argument (self)
print("object ferrari")
ferrari.show()  # car.show(ferrari)

object audi
Model is audi a4
color is blue
object ferrari
Model is ferrari 488
color is green


In [210]:
class car(): 
      
    # init method which is the constructor 
    def __init__(self, model, color): 
        self.model = model 
        self.color = color 
          
    def show(good): # self is not a keyword, we can mention it with any words. It is used to, just represent the instance of this class
        print("Model is", good.model ) 
        print("color is", good.color )

# audi and, ferrari are the Instances of the class car
audi = car("audi a4", "blue")  
# this call is converted to car.__init__(audi, "audi a4", "blue")
# Because of this convertion, we need to define the __init__ method with three arguments which is (self, model, color).
#In the place of self, the instance will be replaced
ferrari = car("ferrari 488", "green") 
# For this instance,
# car.__init__(ferrari, "ferrari 488", "green")


print("object audi")
audi.show()     
# car.show(audi)
# In this call the audi instance is passed as argument, this is why we defined the show method in class car with one argument (self)
print("object ferrari")
ferrari.show()  # car.show(ferrari)

object audi
Model is audi a4
color is blue
object ferrari
Model is ferrari 488
color is green


In [211]:
variable = "good"

class paint():
    var = "brush"
    print("Inside class - ", variable)
    print("Inside class - ", var)
    def class_method(self):
        good = "better"
        print("Inside class_method - ", variable)
        print("Inside class_method - ", self.var)
        print(self.var)

obj = paint()
obj.class_method()

# We can access the variable(defined outside the calss) in the class
# But we can not access the variable(good) defined in the class method from outside class
# See the below table for accessing variables

Inside class -  good
Inside class -  brush
Inside class_method -  good
Inside class_method -  brush
brush


In [212]:
obj.var

'brush'

![](Capture28.png)

###### Inheritance

In [213]:
class parent(object):
    # Constructor 
    def __init__(self, name):
        self.name = name 
    
    def getName(self): 
        return self.name 

    def boolean(self): 
        return False
   
   
# Inherited or Sub class
class child(parent): 
    # Here we return true 
    def boolean(self): 
        return True
    
    emp = parent("Geek1")
    print(emp.getName(), emp.boolean()) 

emp = child("Geek2")
print(emp.getName(), emp.boolean())

Geek1 False
Geek2 True


In [214]:
class parent( object ):     
    
    # __init__ is known as the constructor          
    def __init__(self, name, address):    
        self.name = name 
        self.address = address 
    def display(self):
        print(self.name) 
        print(self.address)
        print("parent class")
        
class child( parent ):            
        def __init__(self, name, address, age, height): 
                self.name = name 
                self.address = address
                self.age = age 
                self.height = height 
  
  
                  
# creation of an object variable or an instance 
a = child('Ram', "India", 3, 50)    
  
a.display()  


Ram
India
parent class
